### DISCUSSION TITLE
Relaunching application on SPUUpdater checkForUpdatesInBackground

### DISCUSSION THREAD

**Discussion by mithril52** [sparkle-project/Sparkle#2458] [Q&A] [answered]
I am having some trouble getting Sparkle to relaunch my application as a part of calling checkForUpdatesInBackground. It will check for updates, download them and install them properly on quit when I manually quit the application. But I can't get it to do the relaunch automatically. None of the below delegate methods seem to be getting called when using checkForUpdatesInBackground:

- updaterShouldRelaunchApplication
- updater:willInstallUpdateOnQuit:immediateInstallationInvocation
- updater:shouldPostponeRelaunchForUpdate:untilInvokingBlock

I do note, however, that updaterShouldRelaunchApplication DOES fire off when using checkForUpdates. Am I perhaps missing something? If I use SUUpdater, the delegate methods above DO fire off, but using this is deprecated so I am trying to move the new stuff. SUUpdater does appear to work when using checkForUpdatesInBackground, but when it tries to relaunch the application, an unhandled exception is thrown. I have been assuming that's because I need to move to the newer classes, but can't seem to get them to work for my particular scenario, so I seem to be stuck. Any help would be greatly appreciated!
 
I have attempted to use all manner of combinations of the configuration settings that potentially relate to this, but no luck. Settings I have tried:

- SUAutomaticallyUpdate
- SUAllowsAutomaticUpdates
- SUEnableAutomaticChecks

Note: I did do some searching through previous issues and discussions, but all the answers I found were for SUUpdater, from 2018 and before.

**Comment by zorgiepoo** [comment, marked as answer]
> updaterShouldRelaunchApplication:

As the API docs state, used for preventing the user from relaunching, so not applicable to you.

> updater:willInstallUpdateOnQuit:immediateInstallationInvocation

Deprecated/old API. Replaced by [-updater:willInstallUpdateOnQuit:immediateInstallationBlock:](https://sparkle-project.org/documentation/api-reference/Protocols/SPUUpdaterDelegate.html#/c:objc(pl)SPUUpdaterDelegate(im)updater:willInstallUpdateOnQuit:immediateInstallationBlock:) which is what you want.

> updater:shouldPostponeRelaunchForUpdate:untilInvokingBlock

For postponing the user relaunching the app, not what you want.

>  SUUpdater does appear to work when using checkForUpdatesInBackground, but when it tries to relaunch the application, an unhandled exception is thrown.

Not familiar with the exception. Please paste it.

Yes you want SUAutomaticallyUpdate and maybe SUEnableAutomaticChecks (if you want to bypass asking user permission for Sparkle checking for updates automatically) and SUAllowsAutomaticUpdates (depending what you set SUEnableAutomaticChecks to). Read the documentation [on the keys](https://sparkle-project.org/documentation/customization/).

**Reply by mithril52** [reply]
Thank you so much, that worked perfectly!
